### PR TITLE
Add typehints to much of the core code.

### PR DESCRIPTION
**Pull request by stephenpardy**
Partially addresses #368 

Adds partial type hints throughout the code. 
Leaves almost all logic untouched. One exception are areas where the default argument was a list. This has been changed to `None` with the list added in the function body. This is to avoid the mutable default argument issue.

---

## What
  * Adds type hints throughout the code

## How to Test
  * Relying on existing unit tests.
  * Can run `mypy` - will still show around 100 issues.

### CODE REVIEW

File: rubicon_ml/client/base.py [hunk 1]
```diff
@@ -9,19 +19,22 @@ class Base:
         The config, which injects the repository to use.
     """
 
-    def __init__(self, domain, config=None):
+    def __init__(self, domain: DOMAIN_TYPES, config: Optional[Config] = None):
```
Comment (ryanSoley) [inline rubicon_ml/client/base.py:22]: I have to look into why the `config` is optional - that doesn't seem right

File: rubicon_ml/client/config.py [hunk 2]
```diff
@@ -29,18 +35,22 @@ class Config:
     """
 
     PERSISTENCE_TYPES = ["filesystem", "memory"]
-    REPOSITORIES = {
+    REPOSITORIES: Dict[str, Type[BaseRepository]] = {
```
Comment (ryanSoley) [inline rubicon_ml/client/config.py:38]: why do we need the `Type` here?

File: rubicon_ml/client/mixin.py [hunk 3]
```diff
@@ -47,14 +60,14 @@ def _validate_data(self, data_bytes, data_file, data_object, data_path, name):
     @failsafe
     def log_artifact(
         self,
-        data_bytes=None,
+        data_bytes: Optional[bytes] = None,
         data_file=None,
```
Comment (ryanSoley) [inline rubicon_ml/client/mixin.py:64]: missed one?

File: rubicon_ml/client/mixin.py [hunk 4]
```diff
@@ -298,8 +316,12 @@ def delete_artifacts(self, ids):
 class DataframeMixin:
     """Adds dataframe support to a client object."""
 
+    _domain: DOMAIN_TYPES
+
     @failsafe
-    def log_dataframe(self, df, description=None, name=None, tags=[]):
+    def log_dataframe(
+        self, df: Union[pd.DataFrame, dd.DataFrame], description=None, name=None, tags=[]
```
Comment (ryanSoley) [inline rubicon_ml/client/mixin.py:323]: missed the last couple

File: rubicon_ml/client/parameter.py [hunk 5]
```diff
@@ -22,36 +31,36 @@ class Parameter(Base, TagMixin):
...
 
     @property
-    def name(self):
+    def name(self) -> Optional[str]:
         """Get the parameter's name."""
         return self._domain.name
 
     @property
-    def value(self):
+    def value(self) -> Optional[Union[object, float]]:
```
Comment (ryanSoley) [inline rubicon_ml/client/parameter.py:49]: isn't `Union[object, float]` == `object`?

File: rubicon_ml/client/parameter.py [hunk 6]
```diff
@@ -22,36 +31,36 @@ class Parameter(Base, TagMixin):
...
         """Get the parameter's value."""
-        return self._domain.value
+        return getattr(self._domain, "value", None)
```
Comment (ryanSoley) [inline rubicon_ml/client/parameter.py:51]: why is this necessary? a parameter domain should never not have a `value` key

File: rubicon_ml/client/rubicon.py [hunk 7]
```diff
@@ -32,7 +33,11 @@ class Rubicon:
     """
 
     def __init__(
-        self, persistence="filesystem", root_dir=None, auto_git_enabled=False, **storage_options
+        self,
+        persistence: Optional[str] = "filesystem",
+        root_dir=None,
+        auto_git_enabled=False,
```
Comment (ryanSoley) [inline rubicon_ml/client/rubicon.py:39]: missed a couple

**Review by ryanSoley — COMMENTED** [review]
this is awesome! thanks for getting us started down this path! 

I have a few general questions and also noted a few functions were you missed a variable or two - lmk if any of those omissions were intentional

File: rubicon_ml/client/base.py (hunk 1 above)
Comment (stephenpardy) [reply rubicon_ml/client/base.py:22]: Yeah, the default value here is `None`, do we need to make it required?

File: rubicon_ml/client/parameter.py (hunk 6 above)
Comment (stephenpardy) [reply rubicon_ml/client/parameter.py:51]: Good catch. This was because I didn't hint the `_domain` value properly.

File: rubicon_ml/client/base.py (hunk 1 above)
Comment (ryanSoley) [reply rubicon_ml/client/base.py:22]: not in this PR, maybe in the future. I think its a remnant of the fact that `Rubicon` used to extend the `Base` and didn't take in a config since it generated it. I'll verify that's the case and if so make an issue to fix it

File: rubicon_ml/client/config.py (hunk 2 above)
Comment (ryanSoley) [reply rubicon_ml/client/config.py:38]: ahh gotcha, makes sense

File: rubicon_ml/client/mixin.py (hunk 4 above)
Comment (stephenpardy) [reply rubicon_ml/client/mixin.py:323]: Thanks.

**Review by ryanSoley — APPROVED** [review]
looks great!
